libcloud/container/drivers/rancher.py

```diff
@@ -0,0 +1,658 @@
+                self.baseuri, env_id), method='POST'
+        )
+        return result.status in VALID_RESPONSE_CODES
+
+    def ex_list_services(self):
+        """
+        List all Rancher Services
+
+        http://docs.rancher.com/rancher/v1.2/en/api/api-resources/service/
+
+        :rtype: ``list`` of ``dict``
+        """
+
+        result = self.connection.request("%s/services" % self.baseuri).object
+        return result['data']
+
+    def ex_deploy_service(self, name, image, environmentid,
+                          start=True, assignserviceipaddress=None,
+                          service_description=None, externalid=None,
+                          metadata=None, retainip=None, scale=None,
+                          scalepolicy=None, secondarylaunchconfigs=None,
+                          selectorcontainer=None, selectorlink=None,
+                          vip=None, **kwargs):
+        """
+        Deploy a Rancher Service under a stack.
+
+        http://docs.rancher.com/rancher/v1.2/en/api/api-resources/service/#create
+
+        *Any further configuration passed applies to the ``launchConfig``*
+
+        :param name: The desired name of the service. (required)
+        :type name: ``str``
+
+        :param image: The Image object to deploy. (required)
+        :type image: :class:`libcloud.container.base.ContainerImage`
+
+        :param environmentid: The stack ID this service is tied to. (required)
+        :type environmentid: ``str``
+
+        :param start: Whether to start the service on creation.
+        :type start: ``bool``
+
+        :param assignserviceipaddress: The IP address to assign the service.
+        :type assignserviceipaddress: ``bool``
+
+        :param service_description: The service description.
+        :type service_description: ``str``
+
+        :param externalid: The externalId for this service.
+        :type externalid: ``str``
+
+        :param metadata: K/V Metadata for this service.
+        :type metadata: ``dict``
+
+        :param retainip: Whether this service should retain its IP.
+        :type retainip: ``bool``
+
+        :param scale: The scale of containers in this service.
+        :type scale: ``int``
+
+        :param scalepolicy: The scaling policy for this service.
+        :type scalepolicy: ``dict``
+
+        :param secondarylaunchconfigs: Secondary container launch configs.
+        :type secondarylaunchconfigs: ``list``
+
+        :param selectorcontainer: The selectorContainer for this service.
+        :type selectorcontainer: ``str``
+
+        :param selectorlink: The selectorLink for this service.
+        :type selectorlink: ``type``
+
+        :param vip: The VIP to assign to this service.
+        :type vip: ``str``
+
+        :return: The newly created service.
+        :rtype: ``dict``
+        """
+
+        launchconfig = {
+            "imageUuid": self._degen_image(image),
+            **kwargs
```

Comment:
So, it does actually work (something new with python3 i guess). What I'm trying to do there is take in extra container options and adding them to the `launchConfig`. This makes it super easy to use that function and for other programs that utilize this (they don't have to worry about building a separate dict if they just want to modify a few options)